### PR TITLE
Fail on ambiguous step definition

### LINKED ISSUE #5864: Ambiguous and missing step definitions
I recently ran into a surprising issue with running Gherkin scenarios with Codeception. Codeception would not run my step definition and after some debugging I figured out that there was another step definition whose pattern matched the current step. When using Behat I would get an exception at that point saying that the step was ambiguous; Behat can't decide which step definition to run. Codeception however just runs the first step definition with a matching pattern (https://github.com/Codeception/Codeception/blob/4.0/src/Codeception/Test/Gherkin.php#L129). I think it would be very useful if it would also throw an exception in that situation. What are your thoughts? I'd be happy to provide a PR for it.

There's one downside: we would have to try all known step definitions to figure out how many of them match, e.g.

```php
$matches = [];
$foundMatchingStepDefinition = false;

foreach ($this->steps as $pattern => $context) {
    $matches = [];
    if (!preg_match($pattern, $stepText, $matches)) {
        continue;
    }
    if ($foundMatchingStepDefinition) {
        throw new AmbiguousStep('...'); // this one is new
    }
    $foundMatchingStepDefinition = true;
    array_shift($matches);
    if ($hasPyStringArg) {
        // get rid off last fake argument
        array_pop($matches);
    }
    if ($stepNode->hasArguments()) {
        $matches = array_merge($matches, $stepNode->getArguments());
    }
}
if (!$foundMatchingStepDefinition) {
    throw new MissingStepDefinition('...'); // this one too
}
call_user_func_array($context, $matches); // execute the step
```

Another issue that we can solve here is for missing step definitions to fail hard instead of showing "I" for "incomplete". I think it's more appropriate.

### PR DESCRIPTION
**Pull request by matthiasnoback**
This fixes #5864

I've decided to hook into the existing step validation code and not throw an exception but use the already present behavior of marking a step as incomplete.

### CODE REVIEW

File: src/Codeception/Test/Gherkin.php [hunk 1]
```diff
@@ -95,15 +95,48 @@ protected function validateStep(StepNode $stepNode)
...
+            // There were no matches, meaning that the user should first add a step definition for this step
+            $incomplete = $this->getMetadata()->getIncomplete();
+            $this->getMetadata()->setIncomplete("$incomplete\nStep definition for `$stepText` not found in contexts");
+        }
+        if (count($matches) > 1) {
+            // There were more than one match, meaning that we don't know which step definition to execute for this step
+            $incomplete = $this->getMetadata()->getIncomplete();
+            $matchingDefinitions = [];
+            foreach ($matches as $pattern => $context) {
+                $matchingDefinitions[] = '- ' . $pattern . ' (' . self::contextAsString($context) . ')';
```
Comment (SamMousa) [inline src/Codeception/Test/Gherkin.php:120]: Should this be a non-static call (or should the helper function be static?)

File: src/Codeception/Test/Gherkin.php [hunk 2]
```diff
@@ -95,15 +95,48 @@ protected function validateStep(StepNode $stepNode)
...
             $res = preg_match($pattern, $stepText);
             if (!$res) {
                 continue;
             }
+            $matches[$pattern] = $context;
+        }
+        if (count($matches) === 1) {
+            // Exactly one step definition matches the given step
             return;
         }
```
Comment (SamMousa) [inline src/Codeception/Test/Gherkin.php:109]: This doesn't add anything, right? 
```suggestion
```

**Review by DavertMik — APPROVED** [review]

**Comment by DavertMik** [comment]
Thank you @matthiasnoback

**Comment by matthiasnoback** [comment]
Cool, thanks!

**Pull request merged**